### PR TITLE
Feature/stock allocation mode

### PR DESCRIPTION
**Pull request by FritzViljoen**
CHANGELOG entry:

Added
Stock allocation modes to loads
How to test this code via the UI:

pallet allocation behaviours
For all modes below: shipping and truck arrival needs to apply same checks/rules, except for the pre-stock and failed_stock modes

1. **stock**(standard): pallets can only be allocated or shipped if in stock (pallet.stock is true)
1. **pre_stock**: pallets can be allocated regardless of stock state or inspected state, but not when inspected and inspection failed
1. **shipping/truck arrival**: at this point pallets must be in stock (inspected state does not matter)
1. **failed_stock**: pallets can be allocated regardless of stock state or inspected state or inspection results
1. **shipping/truck arrival**: pallets must be inspected, but results does not matter
1. **uninspected_stock**: pallets can be allocated regardless of inspected or stock state; same applies to shipping
1. **rmt bins**: pallets can be allocated regardless of stock or inspected state or results, but pallets must have a rmt grade and bin pack-type; same applies to shipping

### CODE REVIEW

File: lib/finished_goods/repositories/load_repo.rb [hunk 1]
```diff
@@ -194,7 +200,11 @@ def list_pallets_for_load(id) # rubocop:disable Metrics/AbcSize
            .where(params.compact)
            .where(Sequel.lit('load_id is null OR load_id = ?', id))
 
-      ds = ds.where(in_stock: true) unless load.depot_return
+      case load_stock_allocation_mode(id)
+      when :standard
```
Comment (jcsjcs) [inline lib/finished_goods/repositories/load_repo.rb:204]: Use a const here.

File: lib/finished_goods/task_permission_checks/load.rb [hunk 2]
```diff
@@ -62,26 +65,101 @@ def delete_check
         all_ok
       end
 
+      def can_allocate_pre_load_check
+        checks = %i[not_scrapped not_on_any_load not_shipped order_spec not_wip
+                    not_failed_otmc has_nett_weight has_gross_weight in_stock
+                    passed_govt_inspection_if_inspected]
+        checks -= %i[has_gross_weight] unless AppConst::CR_PROD.are_pallets_weighed?
+
+        checks -= if AppConst::CR_FG.load_stock_allocation_modes.include? :pre_stock
```
Comment (jcsjcs) [inline lib/finished_goods/task_permission_checks/load.rb:74]: Surely this should depend on the actual mode of the load, not just that it is possible to choose this mode? Or does pre-load mean that this is before there is a load - in which case when is this check used?

File: lib/finished_goods/task_permission_checks/load.rb [hunk 3]
```diff
@@ -62,26 +65,101 @@ def delete_check
...
+        failed_response(e.message)
+      end
+
+      def allocate_pallets_check # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity
+        checks = %i[not_scrapped not_on_load not_shipped order_spec not_wip has_nett_weight has_gross_weight]
+        checks -= %i[has_gross_weight has_nett_weight] if truck_must_be_weighed?
+        checks -= %i[has_gross_weight] unless AppConst::CR_PROD.are_pallets_weighed?
+
+        case @entity.stock_allocation_mode&.to_sym
+        when :standard
```
Comment (jcsjcs) [inline lib/finished_goods/task_permission_checks/load.rb:112]: Use consts.

File: lib/finished_goods/task_permission_checks/load.rb [hunk 4]
```diff
@@ -62,26 +65,101 @@ def delete_check
...
+        truck_arrival_pallets_check
+      end
+
+      def truck_arrival_pallets_check # rubocop:disable Metrics/AbcSize, Metrics/CyclomaticComplexity
+        checks = %i[not_scrapped not_on_load not_shipped order_spec not_wip has_nett_weight has_gross_weight]
+        checks -= %i[has_gross_weight has_nett_weight] if truck_must_be_weighed?
+        checks -= %i[has_gross_weight] unless AppConst::CR_PROD.are_pallets_weighed?
+
+        case @entity.stock_allocation_mode&.to_sym
+        when :standard
```
Comment (jcsjcs) [inline lib/finished_goods/task_permission_checks/load.rb:151]: Consts.

**Review by jcsjcs — CHANGES_REQUESTED** [review]
Please note there is a merge conflict on this PR.

File: lib/finished_goods/task_permission_checks/load.rb (hunk 2 above)
Comment (FritzViljoen) [reply lib/finished_goods/task_permission_checks/load.rb:74]: it is only used in nsli_in.rb

File: lib/client_rules/client_fg_rules.rb [hunk 5]
```diff
@@ -57,9 +56,8 @@ class ClientFgRules < BaseClientRules
             ca_treatment_for_bins: {},
             set_local_stock_via_pi: false,
             wo_fulfillment_pallet_warning_level: 2,
-            enable_depot_returns: false,
             temp_tail_required_to_ship: false,
-            can_allocate_pre_stock_pallets: false },
+            load_stock_allocation_modes: %i[standard] },
```
[the remark this answers is not in the document]
Comment (FritzViljoen) [reply lib/client_rules/client_fg_rules.rb:58]: App_Const is not accessible from within the CLIENT_SETTINGS hash

**Pull request merged**